packages/settings/src/settings-feature-account-generate-vanity.tsx

```diff
@@ -0,0 +1,263 @@
+import { ellipsify } from '@workspace/ui/lib/ellipsify'
+import { handleCopyText } from '@workspace/ui/lib/handle-copy-text'
+import { toastError } from '@workspace/ui/lib/toast-error'
+import { toastSuccess } from '@workspace/ui/lib/toast-success'
+import { useEffect, useReducer, useRef, useState } from 'react'
+import { useNavigate, useParams } from 'react-router'
+
+import {
+  SettingsUiWalletFormGenerateVanity,
+  type VanityWalletFormFields,
+} from './ui/settings-ui-wallet-form-generate-vanity.tsx'
+import { SettingsUiWalletItem } from './ui/settings-ui-wallet-item.tsx'
+
+type VanityGeneratorState =
+  | { attempts: number; error: string | null; result: null; status: 'idle' | 'pending' | 'error' }
+  | { attempts: number; error: null; result: { address: string; secretKey: string }; status: 'success' }
+
+type VanityGeneratorAction =
+  | { type: 'start' }
+  | { payload: number; type: 'progress' }
+  | { payload: string; type: 'error' }
+  | { payload: { address: string; attempts: number; secretKey: string }; type: 'success' }
+  | { type: 'reset' }
+
+const initialVanityGeneratorState: VanityGeneratorState = { attempts: 0, error: null, result: null, status: 'idle' }
+
+function vanityGeneratorReducer(state: VanityGeneratorState, action: VanityGeneratorAction): VanityGeneratorState {
+  switch (action.type) {
+    case 'start':
+      return { attempts: 0, error: null, result: null, status: 'pending' }
+    case 'progress':
+      return { ...state, attempts: action.payload }
+    case 'success':
+      return {
+        attempts: action.payload.attempts,
+        error: null,
+        result: { address: action.payload.address, secretKey: action.payload.secretKey },
+        status: 'success',
+      }
+    case 'error':
+      return { attempts: 0, error: action.payload, result: null, status: 'error' }
+    case 'reset':
+      return initialVanityGeneratorState
+    default:
+      return state
+  }
+}
+
+function useVanityGenerator() {
+  const workerRef = useRef<Worker | null>(null)
+  const [state, dispatch] = useReducer(vanityGeneratorReducer, initialVanityGeneratorState)
+
+  useEffect(() => {
+    return () => {
+      workerRef.current?.terminate()
+      workerRef.current = null
+    }
+  }, [])
+
+  const start = async (input: VanityWalletFormFields) => {
+    const sanitizedPrefix = input.prefix?.trim().slice(0, 4) ?? ''
+    const sanitizedSuffix = input.suffix?.trim().slice(0, 4) ?? ''
+    const payload = {
+      caseSensitive: input.caseSensitive ?? true,
+      prefix: sanitizedPrefix,
+      suffix: sanitizedSuffix,
+    }
+
+    dispatch({ type: 'start' })
+    workerRef.current?.terminate()
+    const worker = new Worker(new URL('./workers/vanity-worker.ts', import.meta.url), { type: 'module' })
+    workerRef.current = worker
+
+    worker.onmessage = (event) => {
+      const { type, payload } = event.data ?? {}
+      if (type === 'progress' && typeof payload === 'number') {
```

Comment:
Can we use type inference here and throughout this file? We should not need to check the type of payload when we check the value of type